File: PlaywireObjC/adtypes/AdTypesViewController.m
//
//  AdTypesViewController.m
//  PlaywireDemo
//
//  Created by Playwire Mobile Team on 01/05/23.
//  Copyright © 2023 Playwire. All rights reserved.
//

#import "AdTypesViewController.h"
#import <Playwire/Playwire-Swift.h>
#import "../ads/view/banner/BannerViewController.h"
#import "../ads/fullscreen/interstitial/InterstitialViewController.h"
#import "../ads/fullscreen/rewarded/RewardedViewController.h"
#import "../ads/fullscreen/appopenad/AppOpenAdViewController.h"
#import "../ads/view/native/NativeAdViewController.h"

typedef NS_ENUM(NSInteger, AdMode) {
    AdModeBanner,
    AdModeInterstitial,
    AdModeRewarded,
    AdModeAppOpen,
    AdModeNative,
};

@interface AdUnitItem: NSObject
@property (nonatomic, copy, readonly) NSString *alias;
@property (nonatomic, assign, readonly) AdMode mode;

- (instancetype)initWithAlias:(NSString *)alias mode:(AdMode)mode;
@end

@implementation AdUnitItem

- (instancetype)initWithAlias:(NSString *)alias mode:(AdMode)mode {
    self = [super init];
    if (self) {
        _alias = [alias copy];
        _mode = mode;
    }
    return self;
}

@end

@interface AdTypesViewController()
@property (strong, nonatomic) NSArray<AdUnitItem *> *adUnitItems;
@property (strong, nonatomic) NSString *cellId;
@end

@implementation AdTypesViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.view.backgroundColor = [UIColor whiteColor];
    self.cellId = @"BasicCell";
    self.adUnitItems = @[
        [[AdUnitItem alloc] initWithAlias:@"banner-320x50-gam" mode:AdModeBanner],
        [[AdUnitItem alloc] initWithAlias:@"banner-320x50-max" mode:AdModeBanner],
        [[AdUnitItem alloc] initWithAlias:@"banner-300x250-gam" mode:AdModeBanner],
        [[AdUnitItem alloc] initWithAlias:@"banner-300x250-max" mode:AdModeBanner],
        [[AdUnitItem alloc] initWithAlias:@"native-gam" mode:AdModeNative],
        [[AdUnitItem alloc] initWithAlias:@"native-max" mode:AdModeNative],
        [[AdUnitItem alloc] initWithAlias:@"app-open-gam" mode:AdModeAppOpen],
        [[AdUnitItem alloc] initWithAlias:@"app-open-max" mode:AdModeAppOpen],
        [[AdUnitItem alloc] initWithAlias:@"interstitial-gam" mode:AdModeInterstitial],
        [[AdUnitItem alloc] initWithAlias:@"interstitial-max" mode:AdModeInterstitial],
        [[AdUnitItem alloc] initWithAlias:@"rewarded-gam" mode:AdModeRewarded],
        [[AdUnitItem alloc] initWithAlias:@"rewarded-video-max" mode:AdModeRewarded],
    ];
    
    self.title = @"Playwire Demo";
    self.navigationController.navigationItem.title = @"Playwire Demo";
    
    [self.tableView registerClass:[UITableViewCell class] forCellReuseIdentifier:self.cellId];
    
    [self setupTableView];

    __weak typeof(self) wself = self;
    
    // Start Playwire SDK with `publisherId` and `appId`.
    // Make sure you run SDK initialization only once.
    [PlaywireSDK.shared startWithPublisherId:@"1024407"
                                       appId:@"702"
                              viewController:self
                                  completion:^(BOOL success, NSError * _Nullable error) {
        if (success) {
            [wself showAdUnits];
        } else {
            [wself showInitializationError:error];
        }
    }];
}

- (void)showAdUnits {
    self.tableView.backgroundView = nil;
    [self.tableView reloadData];
}

- (void)setupTableView {
    UILabel *statusLabel = [UILabel new];
    statusLabel.textAlignment = NSTextAlignmentCenter;
    statusLabel.numberOfLines = 0;
    statusLabel.text = @"⏳ SDK start..";
    self.tableView.backgroundView = statusLabel;
}

- (void)showInitializationError:(NSError * _Nullable)error {
    UILabel *statusLabel = [UILabel new];
    statusLabel.textAlignment = NSTextAlignmentCenter;
    statusLabel.numberOfLines = 0;
    statusLabel.text = [NSString stringWithFormat:@"SDK failed to start.\n%@",
                        error.localizedDescription ?: @"Unknown error"];
    self.tableView.backgroundView = statusLabel;
}

- (nonnull UITableViewCell *)tableView:(nonnull UITableView *)tableView cellForRowAtIndexPath:(nonnull NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:self.cellId forIndexPath:indexPath];
    AdUnitItem *adUnitItem = self.adUnitItems[indexPath.row];
    
    cell.textLabel.text = adUnitItem.alias;
    
    return cell;
}

- (NSInteger)tableView:(nonnull UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.adUnitItems.count;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    AdUnitItem *adUnitItem = [self.adUnitItems objectAtIndex:indexPath.row];
    [self presentViewControllerForAdUnitItem:adUnitItem];
}

- (void)presentViewControllerForAdUnitItem:(AdUnitItem *)adUnitItem {
    UIViewController *viewController = nil;
    
    switch (adUnitItem.mode) {
        case AdModeBanner:
            viewController = [[BannerViewController alloc] initWithAdUnitName:adUnitItem.alias];
            break;
        case AdModeInterstitial:
            viewController = [[InterstitialViewController alloc] initWithAdUnitName:adUnitItem.alias];
            break;
        case AdModeRewarded:
            viewController = [[RewardedViewController alloc] initWithAdUnitName:adUnitItem.alias];
            break;
        case AdModeAppOpen:
            viewController = [[AppOpenAdViewController alloc] initWithAdUnitName:adUnitItem.alias];
            break;
        case AdModeNative:
            viewController = [[NativeAdViewController alloc] initWithAdUnitName:adUnitItem.alias];
            break;
    }
    
    if (viewController) {
        [self.navigationController pushViewController:viewController animated:YES];
    }
}

@end

File: PlaywireSwift/AdTypes/AdTypesViewController.swift
//
//  AdTypesViewController.swift
//  PlaywireDemo
//
//  Created by Playwire Mobile Team on 12/20/22.
//  Copyright © 2022 Playwire. All rights reserved.
//

import UIKit
import Playwire

enum AdMode {
    case banner
    case interstitial
    case rewarded
    case appOpen
    case native
}

struct AdUnitItem {
    let alias: String
    let mode: AdMode
}

final class AdTypesViewController: UITableViewController {
    private let cellId = "BasicCell"
    
    private let adUnitItems: [AdUnitItem] = [
        AdUnitItem(alias: "banner-320x50-gam", mode: .banner),
        AdUnitItem(alias: "banner-320x50-max", mode: .banner),
        AdUnitItem(alias: "banner-300x250-gam", mode: .banner),
        AdUnitItem(alias: "banner-300x250-max", mode: .banner),
        AdUnitItem(alias: "native-gam", mode: .native),
        AdUnitItem(alias: "native-max", mode: .native),
        AdUnitItem(alias: "app-open-gam", mode: .appOpen),
        AdUnitItem(alias: "app-open-max", mode: .appOpen),
        AdUnitItem(alias: "interstitial-gam", mode: .interstitial),
        AdUnitItem(alias: "interstitial-max", mode: .interstitial),
        AdUnitItem(alias: "rewarded-gam", mode: .rewarded),
        AdUnitItem(alias: "rewarded-video-max", mode: .rewarded),
    ]

    override func viewDidLoad() {
        super.viewDidLoad()

        view.backgroundColor = .white

        title = "Playwire Demo"
        navigationController?.navigationItem.title = "Playwire Demo"
        
        tableView.register(UITableViewCell.self, forCellReuseIdentifier: cellId)
        
        setupTableView()
        
        // Start Playwire SDK with `publisherId` and `appId`.
        // Make sure you run SDK initialization only once.
        PlaywireSDK.shared.start(publisherId: "1024407", appId: "702", viewController: self) { [weak self] success, error in
            guard let self else { return }
                    
            if success {
                self.showAdUnits()
            } else {
                self.showInitializationError(error)
            }
        }
    }
    
    private func showAdUnits() {
        tableView.backgroundView = nil
        tableView.reloadData()
    }
    
    private func setupTableView() {
        let statusLabel = UILabel()
        statusLabel.textAlignment = .center
        statusLabel.numberOfLines = 0
        statusLabel.text = "SDK start.."
        tableView.backgroundView = statusLabel
    }
    
    private func showInitializationError(_ error: Error?) {
        let statusLabel = UILabel()
        statusLabel.textAlignment = .center
        statusLabel.numberOfLines = 0
        statusLabel.text = "SDK failed to start.\n\(error?.localizedDescription ?? "Unknown error")"
        tableView.backgroundView = statusLabel
    }
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        adUnitItems.count
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: cellId, for: indexPath)
        
        let adUnitItem = adUnitItems[indexPath.row]
        cell.textLabel?.text = adUnitItem.alias
        
        return cell
    }
    
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        let adUnitItem = adUnitItems[indexPath.row]
        presentViewController(for: adUnitItem)
    }
    
    private func presentViewController(for adUnitItem: AdUnitItem) {
        switch adUnitItem.mode {
        case .banner:
            let bannerVC = BannerViewController(adUnitName: adUnitItem.alias)
            navigationController?.pushViewController(bannerVC, animated: true)
            
        case .interstitial:
            let interstitialVC = InterstitialViewController(adUnitName: adUnitItem.alias)
            navigationController?.pushViewController(interstitialVC, animated: true)
            
        case .rewarded:
            let rewardedVC = RewardedViewController(adUnitName: adUnitItem.alias)
            navigationController?.pushViewController(rewardedVC, animated: true)
            
        case .appOpen:
            let appOpenVC = AppOpenViewController(adUnitName: adUnitItem.alias)
            navigationController?.pushViewController(appOpenVC, animated: true)
            
        case .native:
            let nativeVC = NativeViewController(adUnitName: adUnitItem.alias)
            navigationController?.pushViewController(nativeVC, animated: true)
        }

    }
}
